src/detect-tls-cipher-suite.c

```diff
@@ -0,0 +1,1093 @@
+    {0xC0AE, "TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8"},
+    {0xC0AF, "TLS_ECDHE_ECDSA_WITH_AES_256_CCM_8"},
+    {0xCCA8, "TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256"},
+    {0xCCA9, "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256"},
+    {0xCCAA, "TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256"},
+    {0xCCAB, "TLS_PSK_WITH_CHACHA20_POLY1305_SHA256"},
+    {0xCCAC, "TLS_ECDHE_PSK_WITH_CHACHA20_POLY1305_SHA256"},
+    {0xCCAD, "TLS_DHE_PSK_WITH_CHACHA20_POLY1305_SHA256"},
+    {0xCCAE, "TLS_RSA_PSK_WITH_CHACHA20_POLY1305_SHA256"},
+    {0xFFFF, "" }
+
+};
+
+typedef struct DetectTlsCipherSuiteData_ {
+    uint32_t direction;         /**< STREAM_TOSERVER or STREAM_TOCLIENT */
+    uint32_t length;            /** < Number of cipher suites */
+    uint16_t *ciphersuite;      /** < The lit of cipher suites */
+} DetectTlsCipherSuiteData;
+
+void DetectTlsCipherSuiteFree(void *data)
+{
+    if (data == NULL)
+        return;
+
+    DetectTlsCipherSuiteData *cipher = data;
+    SCFree(cipher);
+    data = NULL;
+}
+
+/**
+ * \brief gets a cipher suite description
+ *
+ * \param value - 2 bytes value for cipher suite code
+ *
+ * \retval "-" if not found
+ * \retval const char * with description if found
+ */
+const char * TlsCipherSuiteDescription(uint16_t value)
+{
+#if __BYTE_ORDER == __BIG_ENDIAN
+    uint16_t I = value;
+#elif __BYTE_ORDER == __LITTLE_ENDIAN
+    uint16_t I = (value >>8) | (value<<8);
+#endif
+    for (size_t i = 0; sizeof(g_suites); ++i) {
+        if (g_suites[i].value == I)
+             return g_suites[i].description;
+        if (g_suites[i].value == 0xFFFF)
+             return g_suites[i].description;
+    }
+    return "-";
+}
+
+/**
+ * \brief match the specified cipher suite for handshake side on a tls session
+ *
+ * \param t pointer to thread vars
+ * \param det_ctx pointer to the pattern matcher thread
+ * \param p pointer to the current packet
+ * \param m pointer to the sigmatch that we will cast into DetectTlsData
+ *
+ * \retval 0 no match
+ * \retval 1 match
+ */
+
+//
+static int DetectTlsCipherSuiteMatch(ThreadVars *t, DetectEngineThreadCtx *det_ctx,
+        Flow *f, uint8_t flags, void *state, void *txv,
+        const Signature *s, const SigMatchCtx *m)
+{
+    SCEnter();
+    DetectTlsCipherSuiteData *data = (DetectTlsCipherSuiteData *)m;
+
+    if (data == NULL) {
+        SCReturnInt(0);
+    }
+    if (! (data->direction & flags)) {
+        SCReturnInt(0);
+    }
+    SSLState *ssl_state = (SSLState *)state;
+    if (ssl_state == NULL) {
+        SCLogDebug("no tls state, no match");
+        SCReturnInt(0);
+    }
+    SSLStateConnp *connp = NULL;
+    if (flags & STREAM_TOSERVER) {
+        connp = &ssl_state->client_connp;
+    } else {
+        connp = &ssl_state->server_connp;
+    }
+
+    if (connp->num_cipher_suites > 0 && data->length > 0) {
+        for (uint32_t j=0; j < data->length; ++j) {
+            for (uint32_t i=0; i < connp->num_cipher_suites; ++i) {
+#if __BYTE_ORDER == __BIG_ENDIAN
```

Comment:
I would prefer having the byte order handling centralized somehow. Too many opportunities for mistakes this way.